src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,422 @@
+    }
+
+    ss::future<errc> add_kv(Key key, Value val) {
+        auto holder = _gate.hold();
+        auto units = co_await _lock.hold_read_lock();
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        iobuf record_key;
+        serde::write(record_key, op_type::kv_data);
+        serde::write(record_key, key);
+        co_return co_await replicate_and_wait(
+          op_type::kv_data, std::move(record_key), offset_data(key, val));
+    }
+
+    ss::future<errc> delete_kv(Key key) {
+        auto holder = _gate.hold();
+        auto units = co_await _lock.hold_read_lock();
+        auto it = kv_data.find(key);
+        if (it == kv_data.end()) {
+            co_return errc::success;
+        }
+        iobuf record_key;
+        // write a tombstone for key
+        serde::write(record_key, op_type::kv_data);
+        serde::write(record_key, key);
+        co_return co_await replicate_and_wait(
+          op_type::kv_data,
+          std::move(record_key),
+          offset_data{key, std::nullopt});
+    }
+
+    ss::future<result<size_t>> repartition(size_t new_partition_count) {
+        auto holder = _gate.hold();
+        if (!_is_routing_partition) {
+            co_return errc::invalid_request;
+        }
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        auto units = co_await _lock.hold_read_lock();
+        auto repartition_units = co_await _repartitioning_lock.get_units();
+        if (_num_partitions && _num_partitions.value() == new_partition_count) {
+            co_return _num_partitions.value();
+        }
+        if (_num_partitions && _num_partitions.value() > new_partition_count) {
+            // reducing the number of partitions is disallowed without
+            // a proper migration protocol.
+            co_return errc::invalid_request;
+        }
+        auto res = co_await replicate_and_wait(
+          op_type::repartitioning,
+          op_type::repartitioning,
+          repartitioning_data(new_partition_count));
+
+        if (res != errc::success) {
+            co_return res;
+        }
+        vassert(_num_partitions.has_value(), "unexpected state");
+        co_return _num_partitions.value();
+    }
+
+private:
+    static constexpr model::partition_id routing_partition{0};
+    static constexpr std::chrono::seconds sync_timeout{5};
+    // a guardrail until an inactivity based expiration is
+    // implemented.
+    static constexpr size_t offset_keys_limit = 10000;
+
+    void do_apply_repartition(iobuf buf) {
+        auto data = serde::from_iobuf<repartitioning_data>(std::move(buf));
+        // todo: add logging on state changes.
+        _num_partitions = data.num_partitions;
+    }
+    void do_apply_assignment(iobuf buf) {
+        auto data = serde::from_iobuf<coordinator_assignment>(std::move(buf));
+        _coordinators[data.key] = data;
+    }
+    void do_apply_offsets(iobuf buf) {
```

Comment:
the whole `kv_stm` seems to be generic but here the name refers to `offsets` i think it should be something like `do_apply_data` ?